[yihuididi] iP

yihuididi opened the pull request:
# YihuiBot

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

A task tracker to manage your todo tasks, YihuiBot is:
- text-based
- easy to learn
- ~~FAST~~ ~~SUPER FAST~~ *EXTREMELY FAST* to use :zap::fire:

To install, simply follow these instructions:
1. Download the `.jar` file from [here](https://github.com/yihuididi/ip/releases/tag/Level-9).
2. Move the `.jar` file to a suitable folder location to run YihuiBot.
3. Depending on which version you are on, you might need to create a folder called `data` on the same level as the `.jar` file.
4. run `java -jar yihuibot.jar` in your command line.

And it is **FREE**!

Features:
- [x] Add and Delete Tasks.
- [x] Mark and Unmark Tasks.
- [x] Filter Tasks by name.
- [ ] Managing Deadlines (coming soon).
- [ ] Reminders (coming soon).

Feel free to change the location to read and store tasks by editing the `main` method in `Yihuibot.java`.
```java
public static void main(String[] args) {
    new YihuiBot("data/task.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yihuididi committed:
Add .class files and vscode files to .gitignore

yihuididi committed:
Change bot name from 'Duke' to 'YihuiBot'
Add a greeting and exit function to bot

yihuididi committed:
YihuiBot.java: Change access modifiers of constant fields
from 'public' to 'private'

The fields 'HORIZONTAL_LINE' and 'NAME' should not be
accessed outside of the class.

yihuididi committed:
YihuiBot.java: create wrapStringWithHorizontalLines method

There will likely be more functions that would require
the print statement to be wrapped in two horizontal lines.
Having this method will simplify things.

YihuiBot.java: Rename 'printGreetings' and 'printExit' to
'greetings' and 'exit' respectively

There will likely bre many more functions like the above two.
Removing the 'print' in the above methods make these methods
less verbose.

yihuididi committed:
YihuiBot.java: move the horizontal line declaration into
the wrapStringWithHorizontalLines method

This is because the horizontal line is only required in
this method.

yihuididi committed:
Improve the skeletal version of YihuiBot so that it
echos commands entered by the user, and exits
when the user types the command 'bye'.

yihuididi committed:
YihuiBot.java: rename 'echo' method to 'addTask'

Let's make the method add tasks into a list instead of
simply echoing the user's input.

yihuididi committed:
Add the functionality for adding tasks

yihuididi committed:
Add the ability for bot to display a list of task

yihuididi committed:
Create a new Task class for all future tasks

yihuididi committed:
Modify YihuiBot.java to use the Task class

yihuididi committed:
Task.java: add a isComplete method

YihuiBot.java: list method now display whether task has been completed

yihuididi committed:
Create a new Message class

This Message class has a print method, which can be called
to print the message in a specific format. In this case, the
format is to wrap the message in horizontal lines.

Let's implement the Message class into YihuiBot so that
the messages print by YihuiBot uses the new Message class.

Remove wrapStringWithHorizontalLines method in YihuiBot
as that method is no longer needed.

yihuididi committed:
Add javadocs for the classes

yihuididi committed:
Change callSuitableFunction method in YihuiBot

The current implementation of main checks if user's input is 'bye'.

Let's extract out this logic to the callSuitableFunction method.

callSuitableFunction now splits the user's input into command and arguments.
It also returns false if user input is 'bye', and true otherwise.

yihuididi committed:
Add the ability to mark tasks as done or undone

yihuididi committed:
Create TodoTask, DeadlineTask, EventTask class

These tasks will inherit the abstract Task class.

Let's implement this new classes into YihuiBot.

yihuididi committed:
Refactor code

Create a new Executable class.

The current implementation of YihuiBot results in too much logic written
in YihuiBot.java.

Let's implement the Executable class. Instance of this class can be
executed by YihuiBot. Calling Executable.execute() will execute it
based on what type of executable it is. Example of Executables are AddTask
and ListTask.

Create a new Parser class.

The parser will parse user's input, and create the appropriate executable
so that it can be executed by YihuiBot.

yihuididi committed:
Modify runtest files and expected output

yihuididi committed:
Add new Exception class

The exception package handles all wrong user input to YihuiBot.

Let's implement it into the parser. The parser has functions to
throw suitable exceptions based on what input was given.

Create new prettyPrint function in YihuiBot. The prettyPrint will
print all output of the bot's executable wrapped in horizontal lines.

Change run test files to follow new implementation.

yihuididi committed:
Track exception package

The previous commit forget to track the exception package.

yihuididi committed:
Add support for deleting tasks from list

Change regression testing testcases and EXPECTED.TXT
to reflect change.

yihuididi committed:
Use enums for Executable exit code

yihuididi committed:
Refactor code

The parser does not throw any exceptions when invalid arguments are given.
It relies on what the parser returns when parsing in a user input. If there
were invalid arguments, it will return a Warning executable. This is
not the best way to do it. It should throw exceptions when the input is not
correct, and it should be catched by the bot.

Let's implement it as such. The parser will now throw an exception due to user
input. YihuiBot will catch these exceptions and act accordingly.

yihuididi committed:
Tasks can be initialized by ./data/task.txt if file exists

Create new IncorrectTaskFormatException when file is corrupted.

TaskReader will parse the file to initialize task according to file.

Modify test expected results to complement change.

yihuididi committed:
Save tasks into data/task.txt

Modify test files to complement this change.

yihuididi committed:
Merge branch 'branch-Level-7'

yihuididi committed:
Store date time as an LocalDateTime object

yihuididi committed:
Merge branch 'branch-Level-8'

yihuididi committed:
Refactor code

The commands in YihuiBot are static.

Users should be able to instantiate the bot whenever they please.

Let's make these methods static.
Refactor YihuiBot to mainly two functions, the constructor and run().
The constructor will instantiate a Ui, Storage, and TaskList class.

Let's create the Ui class.
This class is responsible for all user interface, parsing the user's input
to generate an Executable, as well as formatting the message of YihuiBot
nicely. Contains methods such as prettyPrint and warningPrint. It instantiates
the Parser for generating Executables.

Let's create the Storage class.
This class is responsible for reading the data in specified file and
converting them into a TaskList. It is also responsible for writing the tasks
from TaskList into file. Employs a TaskReader and TaskWriter for each
scenarios.

Let's create the TaskList class. This class extends the ArrayList<Task>
and is used to store the list of tasks created by user. This TaskList would
then be passed to Storage as argument when writing the tasks into file.

Create many new Exceptions, one package for taskformatting exceptions,
one package for parsing exceptions, and one package executable exceptions.

yihuididi committed:
Merge branch 'A-MoreOOP'

yihuididi committed:
Change description of intialising TaskList

yihuididi committed:
Merge branch 'branch-A-MoreOOP'

yihuididi committed:
Add yihuibot package

All packages are subpackages of yihuibot.

yihuididi committed:
Merge branch 'branch-A-Packages'

yihuididi committed:
Merge branch 'add-gradle-support'

yihuididi committed:
Correct a style typo in build.gradle

yihuididi committed:
Add unit test for YihuiBot, Parser, and Ui

yihuididi committed:
Add unit test for Storage, TaskReader, and TaskWriter

yihuididi committed:
Merge branch 'branch-A-JUnit'

yihuididi committed:
Add javadocs to Parser.java

yihuididi committed:
Edit UnmarkTask.java to follow coding standrard

yihuididi committed:
Gives users to find task by searching a keyword

Create a new FilterTask Executable and allow Parser to create this Executable
when the find command is given.

Change testcases to match this new change.

yihuididi committed:
Merge branch 'branch-A-JavaDoc'

yihuididi committed:
Merge branch 'branch-A-CodingStandard'

yihuididi committed:
Merge branch 'branch-Level-9'

yihuididi committed:
Create directory at file path when running bot

yihuididi committed:
Allow users to save tasks to filepath with no directories

The program only allow users to save tasks to filepath nested
in directories. i.e. it will allow filepath data/task.txt
but not filepath task.txt.

This implementation is not flexible.

Let's make it more flexible by changing the constructor method
in storage/Storage.java.

Edit test files to follow this new change.

yihuididi committed:
Add checkstyle to gradle

yihuididi committed:
Change classes in ui package to follow coding standards

yihuididi committed:
Change yihuibot/YihuiBot.java to follow coding standards

yihuididi committed:
Change storage package to follow coding standards

yihuididi committed:
Change executable package to follow coding standards

yihuididi committed:
Change task package to follow coding standards

yihuididi committed:
Change src/test files to follow coding standards

yihuididi committed:
Merge branch 'branch-A-CheckStyle'

yihuididi committed:
Move YihuiBot from a CLI app to a GUI app

Set up dependencies for running JavaFX.
Create a simple application that just greets the user for now.

yihuididi committed:
Create DialogBox for user's and bot's dialogs

yihuididi committed:
Add ability for bot to respond to user

yihuididi committed:
Separate JavaFX controls, styling and nodes

yihuididi committed:
Add CSS to JavaFX components

yihuididi committed:
Bot responds according to user's input

yihuididi committed:
Move JavaFX nodes to yihuibot.component package

The nodes are in the same level as Main.
As the program gets more complicated, and more nodes
are created, the yihuibot package may get overcrowded.

Solve by moving all node components to yihuibot.component.

yihuididi committed:
Track yihuibot.component

yihuididi committed:
Edit automated tests

yihuididi committed:
Merge branch 'branch-Level-10'

yihuididi committed:
Use variable arguments for Parser commands

yihuididi committed:
Merge branch 'branch-A-Varargs'

winstonlimjy commented on src/main/java/yihuibot/executable/UnmarkTask.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class UnmarkTask extends TaskModifier {
+    private String output;
+    private int idx;
```
Should the variable be named idx?

winstonlimjy replied:
Should the variable be named idx?

winstonlimjy reviewed (commented):
Well done! I could not find any obvious coding standard violations!

winstonlimjy reviewed (commented):
Well done! I could not find any obvious coding standard violations! Keep up the good work!

Shux347 commented on src/main/java/yihuibot/executable/DeleteTask.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class DeleteTask extends TaskModifier {
+    private String output;
+    private int idx;
```
idx could be named better as taskIndex. Avoid abbreviations

Shux347 commented on src/main/java/yihuibot/Main.java:
```diff
@@ -0,0 +1,66 @@
+    private YihuiBot yihuiBot;
+
+    /**
+     * Constructor for a new Main, with parameter to specify where to read tasks info from.
```
In method header comments, the first sentence should start in the form Returns ..., Sends ..., Adds ... etc. (not Return or Returning etc.)
```suggestion
     * Constructs a new Main, with parameter to specify where to read tasks info from.
```

Shux347 reviewed (commented):
Overall LGTM. The only naming convention error I could find is the use of idx.

Coding standard wise you will need to change the method summary in your header comments to align with the convention given in our coding standards.

yihuididi committed:
Add assertions to Parser.java

The deadline and event methods will look for index words such as
/by and /to. The index of these words cannot be of certain value.
For example, the index of /by cannot be 0 because there needs to
be description before /by.

Let's assert these criteria as proceeding may lead to bugs if it
turns out that index is an unexpected value.

yihuididi committed:
Add assertions to TaskReader.java

The TaskReader will read the information from the file to determine the
tasks to be added. It is expected to throw a TypeException when it reads
an invalid type from the file. If there was an invalid type, but no
TypeException is thrown, an error will occur.

Let's assert such that upon reading the information from the file, the task
type is valid before proceeding.

yihuididi committed:
Merge pull request #1 from yihuididi/branch-A-Assertions

Branch a assertions

yihuididi committed:
Refactor Parser class

Parser has over 300 lines of code. The logic behind Parser is
solely coded in Parser.java. This could lead to difficulty in
debugging as there are a lot of lines to read.

Extract the logic of Parser class using a new Function package.

Depending on the command given by user, Parser will instantiate
a suitable Function to process the arguments of the user.
Depending on what was typed by the user as arguments, the Function
will return a suitable Executable.

yihuididi committed:
Merge branch 'master' into branch-A-CodeQuality

yihuididi committed:
Merge pull request #2 from yihuididi/branch-A-CodeQuality

Refactor Parser class

yihuididi committed:
Merge branch 'master' into branch-A-Streams

yihuididi committed:
Use Streams to filter tasks

Let Stream methods to do the filtering for FilterTask Executable.

yihuididi committed:
Merge pull request #3 from yihuididi/branch-A-Streams

Branch a streams

yihuididi committed:
Introduce CI using Github Actions

Set up Github Actions to perform workflows whenever there is a code push
or when there is a pull request.

yihuididi committed:
Provide more flexibility with the data source

User can specify which file to use as data source.

Using the yihuibot.jar, they can run 'java -jar yihuibot.jar <filepath>'
to specify the fielpath as data source.
Using gradlew, they can run './gradlew run --args='<filepath>''
to specify the filepath as data source.

If not command line arguments is provided, the program will default to
'data/task.txt'.

yihuididi committed:
Merge pull request #4 from yihuididi/branch-BCD-Extension

Provide more flexibility with the data source

yihuididi committed:
Change interface of GUI

Make the height of each dialog box smaller.

Profile picture of messenger is circular.

Change chat background wallpaper.

yihuididi committed:
Merge branch 'branch-A-BetterGui'

yihuididi committed:
Set stage name to name of bot

yihuididi committed:
Add a representative screenshot of product

yihuididi committed:
Update README.md

yihuididi committed:
Update README.md to include instructions for quick start

yihuididi committed:
Update README.md to follow markdown standards

yihuididi committed:
Update README.md again to follow markdown standards
